### PR TITLE
[MAINTENANCE] TYPO3 v11 compatibility

### LINKED ISSUE #775: Support for new TYPO3 LTS version 11.5
## Description

The upcoming Kitodo.Presentation 4.0 release will support TYPO3 9.5 LTS and 10.4 LTS. The next [TYPO3 LTS version](https://typo3.org/cms/roadmap) is already there:

* TYPO3 11.5 LTS Released: 2021-10-05

The next major release 5.x should support TYPO3 10.4 and 11.5 LTS.

## Expected benefits of this development
* Developer / Maintainer
   * This is a must have topic. We cannot ignore the development of the higher-level TYPO3 CMS. It pushes the development and makes sure Kitodo.Presentation is an up-to-date tool.
* Administrators
   * Run supported TYPO3-version is obligatory.

## Estimated Costs and Complexity

I assume this should be a **mid-range project** for TYPO3 11.5.

## Related Issues

* #512 (prerequisite)

### PR DESCRIPTION
**Pull request by christopher-timm**
This PR ensures that the DLF Extension is also compatible with version 11 of Typo3.
The biggest adjustment was the reorganization of the pagination, since the widgets were removed from Typo3. This required an own "Pagniator" for the Solr, as well as for the display of the "PageGrid"

### CODE REVIEW

**Comment by sebastian-meyer** [comment]
Could you please bring this pull request up to date with master? Thank you!

**Comment by sebastian-meyer** [comment]
Could you please adjust the testing workflow in order to ensure testing against TYPO3 v10 and v11 instead of v9 and v10? Currently tests fail, because they are still run against v9.

**Comment by sebastian-meyer** [comment]
I think we need to update [`Build/Test/runTests.sh`](https://github.com/kitodo/kitodo-presentation/blob/master/Build/Test/runTests.sh) as well. Have a look at the upstream file from TYPO3 CMS: [v10](https://github.com/TYPO3/typo3/blob/10.4/Build/Scripts/runTests.sh), [v11](https://github.com/TYPO3/typo3/blob/11.5/Build/Scripts/runTests.sh).

Constraints for tests and in composer.json for TYPO3 v10 and v11 should be (see also: https://get.typo3.org/):
- PHP: **7.4** (v10 doesn't support PHP 8 and v11 requires at least PHP 7.4)
- MariaDB: **10.3** (default for both TYPO3 versions)
- TYPO3: **10.4.36**|**11.5.24** (latest safe TYPO3 versions)

Also please check if we can use the newest version of [spatie/phpunit-watcher](https://packagist.org/packages/spatie/phpunit-watcher) since we should be able to run PHPUnit 9.x now, are we?

After updating composer.json please also update composer.lock file accordingly (by running `composer update` locally).

**Comment by sebastian-meyer** [comment]
There are a lot of mentions of TYPO3 v9 left in the documentation and code. All of them need to be updated according to the new supported versions. Also, some of them point to work-arounds for backwards compatibility which are no longer needed when dropping support for v9. Please check and remove those work-arounds.

**Comment by csidirop** [comment]
I did a few tests (on TYPO3 v10 with DFG-Viewer) and while most of the parts I tested worked fine I found following error: 

Using the search or collection plugin leads to `Core: Error handler (FE): PHP Warning: A non-numeric value encountered in /var/www/typo3/vendor/typo3fluid/fluid/src/Core/Parser/SyntaxTree/Expression/MathExpressionNode.php line 77`. 

For me this results in some erroneous output: 
```
Die Suche ergab 8 Treffer in 8 Dokumenten.

Einträge -24 bis 0 von 8. 
```

Here what it looks like: (without custom CSS)
grafik
and 
grafik

**Comment by christopher-timm** [comment]
@csidirop @sebastian-meyer : Do you both get the error with the wrong number of hits?
I can't reproduce this on my local dev environment. Can you give me more hints, so I can reproduce the error?
Bildschirmfoto 2023-02-28 um 19 26 06

**Comment by csidirop** [comment]
First of all, could you please bring your branch up to date with kitodo:master as @sebastian-meyer requested one month ago. You are at 31dbff376f0ca8eb72bfe6991264fc490c50c2fc and [38 commits behind](https://github.com/effective-webwork/kitodo-presentation/compare/typo3v11...kitodo:kitodo-presentation:master), some of which fixed bugs.

Now, I get the same results regardless weather I'm merging kitodo:master into your branch or not, its still the same error from above. 

But it may also have something to do with my dev environment. I'll try that asap. But for the moment what I'm doing is checking out your branch in my Docker env after everything is set up (here may be the difference?). Plus I'm using DFG-Viewer which may have some influence. I'll test that too asap.

**Comment by csidirop** [comment]
So I updated my docker setup to initially setup dlf with your branch, but as soon as I want to add Kitodo Elements to a subpage I get this error: 
```
Core: Exception handler (WEB): Uncaught TYPO3 Exception: #1478105826: Data structure file /var/www/typo3/public/typo3conf/ext/dlf/Configuration/FlexForms/PageView.xml could not be resolved to an existing file | RuntimeException thrown in file /var/www/typo3/public/typo3/sysext/core/Classes/Configuration/FlexForm/FlexFormTools.php in line 688. Requested URL: http://localhost/typo3/index.php?route=%%2Frecord%%2Fedit&token=--AnonymizedToken--&edit%%5Btt_content%%5D%%5B12%%5D=new&defVals%%5Btt_content%%5D%%5BcolPos%%5D=0&defVals%%5Btt_content%%5D%%5Bsys_language_uid%%5D=0&returnUrl=%%2Ftypo3%%2Findex.php%%3Froute%%3D%%252Fmodule%%252Fweb%%252Flayout%%26token%%3D95445b61554c1e3a30cd87be3f6451301c9e5fce%%26id%%3D12%%26&defVals[tt_content][CType]=list&defVals[tt_content][list_type]=dlf_pageview 
```

This issue can be resolved by merging kitodo:master. But after that I'm stuck at some config issue I have no time to resolve now.

But I can show you my workflow in case you want to follow this:

1. `git clone https://github.com/UB-Mannheim/kitodo-presentation-docker` and cd into
2. `git checkout dfg-viewer-6.x-ocr` (unfortunately solr is only on our ocr branches available so far)
3. Setup the `.env`-file
4. `docker compose --profile with-solr up [--build]`
5. `docker exec -ti kitodo-presentation-main bash`
6. And than go the dlf folder add your branch and check it out

I can't rule out, that the error only occurs in my environment. Hopefully sebastian can test it and see if he has the same errors.

**Comment by stweil** [comment]
> This required an own "Pagniator" for the Solr

Paginator?

**Comment by christopher-timm** [comment]
> > This required an own "Pagniator" for the Solr
> 
> Paginator?

Typo3 now provides interfaces with which you can implement your own pagination. Two classes are provided in the core, for arrays and for QueryResults. All other types have to be implemented by yourself.
https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Pagination/Index.html

**Comment by sebastian-meyer** [comment]
After some testing, I have found a few more bugs (not all of these are necessarily related to your changes):

- [x] In TYPO3 v11 all the labels in the backend are missing (in Extension Configuration, list module, database records, ...). (In TYPO3 v10 this works fine.)
- [x] Please have a look at the failing tests as well. There are some translation issues and changes to the DB schema which are not properly addressed.
- [x] In the New Tenant BE module some labels are missing in both TYPO3 versions: The message for missing namespaces and the button text for importing the metadata mappings. The latter miraculously appears after clicking "Create namespaces"... (This should be fixed with #926)
- [x] Importing metadata configuration via the New Tenant BE module results in an error (see #916). This works after applying the patch from #917. Nevertheless, we should provide at least a default label for every metadata field and structure entry.
- [x] When indexing new documents, in TYPO3 v11 some metadata are missing in the database (`title`, `title_sorting`, `document_format`, `urn`, `purl`, `license`, `terms`, `restrictions`, and potentially others which were empty anyway for my example records).
- [x] In `Classes/Common/Indexer.php` in line 111 there is a reference to `$this` although the method `add()` is static. Thus indexing in Solr doesn't work. This is related to your changes regarding the initialisation of the `documentRepository`.
- [ ] In the backend module "Upgrade" -> "Check TCA Migrations" there are some migrations listed which have to be performed to ensure full compatibility with TYPO3v11.
- [ ] In the backend module "Upgrade" -> "Scan Extension Files" there are still a lot of warnings about deprecated or even removed methods. Please sort out at least all deprecation messages up to and including TYPO3v10 and all removed methods. (See [API documentation](https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ExtensionArchitecture/HowTo/UpdateExtensions/ExtensionScanner.html) for how to handle false positives.)

I'll continue testing and add more issues if they occur.

**Comment by sebastian-meyer** [comment]
- [x] The labels `newTenant.formatNotOkay` and `newTenant.formatNotOkayMsg` are still missing in v11.
- [x] Some labels for database fields are missing in the backend (in v10 and v11), for example for tables `tx_dlf_libraries` (tab 'Catalogs', `tx_dlf_documents` ('METS Label', 'METS Orderlabel'), `tx_dlf_structures` ('index_name') and maybe others as well.

**Comment by christopher-timm** [comment]
I added the missing labels for the database fields. A faulty search and replace caused "Label" to get kicked out.
The labels "newTenant.formatNotOkay" and "newTenant.formatNotOkayMsg" are present in the language files. But also with my local system, the output remains empty.

**Comment by christopher-timm** [comment]
I think the escaped single quotes cause problems in this case. With double quotes, it works for me.

**Comment by sebastian-meyer** [comment]
I don't know why (because some lines below it works perfectly with single quotes), but now it works for me as well!

File: Resources/Private/Language/locallang_labels.xlf [hunk 1]
```diff
@@ -113,7 +113,7 @@
 			<trans-unit id="tx_dlf_documents.collections">
 				<source>Collections</source>
 			</trans-unit>
-			<trans-unit id="tx_dlf_documents.mets_trans-unit">
+			<trans-unit id="tx_dlf_documents.mets_label">
 				<source>METS @LABEL</source>
 			</trans-unit>
 			<trans-unit id="tx_dlf_documents.mets_ordertrans-unit">
```
Comment (sebastian-meyer) [inline Resources/Private/Language/locallang_labels.xlf:119]: This should be `tx_dlf_documents.mets_orderlabel`.

**Comment by sebastian-meyer** [comment]
And `tx_dlf_structures.index_name` is wrong, too. Currently it is `tx_dlf_structures.id_name`.

**Comment by christopher-timm** [comment]
I have adjusted the TCA accordingly and added information in development.md for v11 and v12.
With this the points should be finished so far. I hope that I have considered everything in the "Scan Extension Files". There were quite a lot of wrong information.

**Comment by csidirop** [comment]
Unfortunately the missing search results are still an issue in my tests. Can you reproduce this?

**Comment by sebastian-meyer** [comment]
I had to resolve a lot of merge conflicts, so maybe there are some issues now. Please extensively test the master branch and compare it to what your intended changes for TYPO3v11 compatibility should have been!

There is still some work to do, but for now this is finally merged!

**Comment by stweil** [comment]
CI tests now fail. Is this related to this PR?

```
PHP Fatal error:  Access level to Kitodo\Dlf\Controller\ToolboxController::setPage() must be protected (as in class Kitodo\Dlf\Controller\AbstractController) or weaker in /home/runner/work/kitodo-presentation/kitodo-presentation/Classes/Controller/ToolboxController.php on line 477
```

**Comment by sebastian-meyer** [comment]
Yes, it most likely is. There probably are more issues than just that...

**Comment by sebastian-meyer** [comment]
OK, that was my fault. I accidently duplicated a method while resolving merge conflicts. This is fixed now.

But tests still don't run through. @chrizzor Could you please have a look into this?

**Comment by csidirop** [comment]
**Tests under TYPO3 v10 with [our DFG-Viewer](https://github.com/UB-Mannheim/dfg-viewer/tree/6.x-ocr):**

- Everything search related is still not working. See my comment from [24 feb](https://github.com/kitodo/kitodo-presentation/pull/885#issuecomment-1443367080). Is it just me?
- The thumbnail preview (gridpage) shows only the first 24 pages and has no option to load more pages: (dfg-viewer issue?)
grafik
Short addition: On an other document the text part was partially linked. First 24 pages with working link for the next 24:
grafik
but the next page had only a working link back:
grafik
The two mentioned mets: https://digi.bib.uni-mannheim.de/fileadmin/vl/ubmaosi/59087/59087.xml and https://digi.bib.uni-mannheim.de/fileadmin/vl/ubmaosi/59088/59088.xml
- In addition, the numbers 1 and 31 refer to the number of review pages. But are somehow incorrectly displayed / placed. For reference, how it currently looks (and should be looking):
grafik

- The image manipulation tools are missing: (dfg-viewer issue?)
grafik

- the same Metadata pairing/grouping issue described later


**Tests under TYPO3 v11:**
- I could not get the viewer running. But that can be on me. I had always issues manually setting presentation up. Ill try more. Is there anywhere a step by step guide?
  -> I could not try the viewer or anything related 
- New Tenant Module:
  - Can not create a new Solr Core:
grafik
    -> I did not set the correct solr configuration prior. But till v10 this wasn't necessary.
  -  After creating namespaces, Structures and Metadata they aren't "grouped" or "paired" by language any more:
   grafik grafik
- Analyze Database Structure shows:
```
Add tables
    CREATE TABLE tx_dlf_documents (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `hidden` SMALLINT DEFAULT 0 NOT NULL, `starttime` INT DEFAULT 0 NOT NULL, `endtime` INT DEFAULT 0 NOT NULL, `fe_group` VARCHAR(100) DEFAULT '' NOT NULL, `prod_id` VARCHAR(255) DEFAULT '' NOT NULL, `location` VARCHAR(255) DEFAULT '' NOT NULL, `record_id` VARCHAR(255) DEFAULT '' NOT NULL, `opac_id` VARCHAR(255) DEFAULT '' NOT NULL, `union_id` VARCHAR(255) DEFAULT '' NOT NULL, `urn` VARCHAR(255) DEFAULT '' NOT NULL, `purl` VARCHAR(255) DEFAULT '' NOT NULL, `title` TEXT NOT NULL, `title_sorting` TEXT NOT NULL, `author` VARCHAR(255) DEFAULT '' NOT NULL, `year` VARCHAR(255) DEFAULT '' NOT NULL, `place` VARCHAR(255) DEFAULT '' NOT NULL, `thumbnail` VARCHAR(255) DEFAULT '' NOT NULL, `structure` INT DEFAULT 0 NOT NULL, `partof` INT DEFAULT 0 NOT NULL, `volume` VARCHAR(255) DEFAULT '' NOT NULL, `volume_sorting` VARCHAR(255) DEFAULT '' NOT NULL, `license` VARCHAR(255) DEFAULT '' NOT NULL, `terms` VARCHAR(255) DEFAULT '' NOT NULL, `restrictions` VARCHAR(255) DEFAULT '' NOT NULL, `out_of_print` TEXT NOT NULL, `rights_info` TEXT NOT NULL, `collections` INT DEFAULT 0 NOT NULL, `mets_label` TEXT NOT NULL, `mets_orderlabel` TEXT NOT NULL, `owner` INT DEFAULT 0 NOT NULL, `solrcore` INT DEFAULT 0 NOT NULL, `status` SMALLINT UNSIGNED DEFAULT 0 NOT NULL, `document_format` VARCHAR(100) DEFAULT '' NOT NULL, INDEX `parent` (pid), INDEX `location` (location), INDEX `record_id` (record_id), INDEX `partof` (partof), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_structures (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_diffsource` MEDIUMBLOB NOT NULL, `hidden` SMALLINT DEFAULT 0 NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `toplevel` SMALLINT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `index_name` VARCHAR(255) DEFAULT '' NOT NULL, `oai_name` VARCHAR(255) DEFAULT '' NOT NULL, `thumbnail` INT DEFAULT 0 NOT NULL, `status` SMALLINT UNSIGNED DEFAULT 0 NOT NULL, INDEX `parent` (pid), INDEX `language` (l18n_parent, sys_language_uid), INDEX `index_name` (index_name), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_metadata (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_diffsource` MEDIUMBLOB NOT NULL, `hidden` SMALLINT DEFAULT 0 NOT NULL, `sorting` INT DEFAULT 0 NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `index_name` VARCHAR(255) DEFAULT '' NOT NULL, `format` INT DEFAULT 0 NOT NULL, `default_value` VARCHAR(255) DEFAULT '' NOT NULL, `wrap` TEXT NOT NULL, `index_tokenized` SMALLINT DEFAULT 0 NOT NULL, `index_stored` SMALLINT DEFAULT 0 NOT NULL, `index_indexed` SMALLINT DEFAULT 0 NOT NULL, `index_boost` DOUBLE PRECISION DEFAULT '1.00' NOT NULL, `is_sortable` SMALLINT DEFAULT 0 NOT NULL, `is_facet` SMALLINT DEFAULT 0 NOT NULL, `is_listed` SMALLINT DEFAULT 0 NOT NULL, `index_autocomplete` SMALLINT DEFAULT 0 NOT NULL, `status` SMALLINT UNSIGNED DEFAULT 0 NOT NULL, INDEX `parent` (pid), INDEX `language` (l18n_parent, sys_language_uid), INDEX `index_name` (index_name), INDEX `index_autocomplete` (index_autocomplete), INDEX `is_sortable` (is_sortable), INDEX `is_facet` (is_facet), INDEX `is_listed` (is_listed), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_metadataformat (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `parent_id` INT DEFAULT 0 NOT NULL, `encoded` INT DEFAULT 0 NOT NULL, `xpath` VARCHAR(1024) DEFAULT '' NOT NULL, `xpath_sorting` VARCHAR(1024) DEFAULT '' NOT NULL, `mandatory` SMALLINT DEFAULT 0 NOT NULL, INDEX `parent` (pid), INDEX `parent_id` (parent_id), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_formats (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `type` VARCHAR(255) DEFAULT '' NOT NULL, `root` VARCHAR(255) DEFAULT '' NOT NULL, `namespace` VARCHAR(255) DEFAULT '' NOT NULL, `class` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_solrcores (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `index_name` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), INDEX `index_name` (index_name), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_collections (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_diffsource` MEDIUMBLOB NOT NULL, `hidden` SMALLINT DEFAULT 0 NOT NULL, `fe_group` VARCHAR(100) DEFAULT '' NOT NULL, `fe_cruser_id` INT DEFAULT 0 NOT NULL, `fe_admin_lock` SMALLINT DEFAULT 0 NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `index_name` VARCHAR(255) DEFAULT '' NOT NULL, `index_search` TEXT NOT NULL, `oai_name` VARCHAR(255) DEFAULT '' NOT NULL, `description` TEXT NOT NULL, `thumbnail` VARCHAR(255) DEFAULT '' NOT NULL, `priority` SMALLINT DEFAULT 3 NOT NULL, `documents` INT DEFAULT 0 NOT NULL, `owner` INT DEFAULT 0 NOT NULL, `status` SMALLINT UNSIGNED DEFAULT 0 NOT NULL, INDEX `parent` (pid), INDEX `language` (l18n_parent, sys_language_uid), INDEX `index_name` (index_name), INDEX `oai_name` (oai_name), INDEX `pid_cruser` (pid, fe_cruser_id), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_libraries (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `cruser_id` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_diffsource` MEDIUMBLOB NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `index_name` VARCHAR(255) DEFAULT '' NOT NULL, `website` VARCHAR(255) DEFAULT '' NOT NULL, `contact` VARCHAR(255) DEFAULT '' NOT NULL, `image` VARCHAR(255) DEFAULT '' NOT NULL, `oai_label` VARCHAR(255) DEFAULT '' NOT NULL, `oai_base` VARCHAR(255) DEFAULT '' NOT NULL, `opac_label` VARCHAR(255) DEFAULT '' NOT NULL, `opac_base` VARCHAR(255) DEFAULT '' NOT NULL, `union_label` VARCHAR(255) DEFAULT '' NOT NULL, `union_base` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), INDEX `language` (l18n_parent, sys_language_uid), INDEX `index_name` (index_name), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_tokens (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL COMMENT 'Timestamp of the token used to determine if it has expired.', `crdate` INT UNSIGNED DEFAULT 0 NOT NULL, `cruser_id` INT UNSIGNED DEFAULT 0 NOT NULL, `token` VARCHAR(255) DEFAULT '' NOT NULL, `options` MEDIUMTEXT NOT NULL, `ident` VARCHAR(30) DEFAULT '' NOT NULL, INDEX `parent` (pid), INDEX `token` (token), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE `tx_dlf_relations` (`uid` INT AUTO_INCREMENT NOT NULL, `uid_local` INT DEFAULT 0 NOT NULL, `uid_foreign` INT DEFAULT 0 NOT NULL, `tablenames` VARCHAR(30) DEFAULT '' NOT NULL, `sorting` INT DEFAULT 0 NOT NULL, `sorting_foreign` INT DEFAULT 0 NOT NULL, `ident` VARCHAR(30) DEFAULT '' NOT NULL COMMENT 'An identifier to describe which tables are matched.', INDEX `local_foreign` (uid_local, uid_foreign, ident), INDEX `uid_local` (uid_local), INDEX `uid_foreign` (uid_foreign), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB COMMENT = 'Pivot table for many-to-many relations between tables. In particular, this is used to match documents and collections by using ident=docs_colls.'
    CREATE TABLE tx_dlf_basket (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `tstamp` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sys_language_uid` INT DEFAULT 0 NOT NULL, `l18n_diffsource` MEDIUMBLOB NOT NULL, `fe_user_id` INT DEFAULT 0 NOT NULL, `l18n_parent` INT DEFAULT 0 NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, `session_id` VARCHAR(32) DEFAULT '' NOT NULL, `doc_ids` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), INDEX `language` (l18n_parent, sys_language_uid), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_printer (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `print` VARCHAR(255) DEFAULT '' NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_mail (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `sorting` INT DEFAULT 0 NOT NULL, `mail` VARCHAR(255) DEFAULT '' NOT NULL, `name` VARCHAR(255) DEFAULT '' NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
    CREATE TABLE tx_dlf_actionlog (`uid` INT AUTO_INCREMENT NOT NULL, `pid` INT DEFAULT 0 NOT NULL, `crdate` INT DEFAULT 0 NOT NULL, `deleted` SMALLINT DEFAULT 0 NOT NULL, `user_id` INT DEFAULT 0 NOT NULL, `file_name` VARCHAR(255) DEFAULT '' NOT NULL, `count_pages` INT DEFAULT 0 NOT NULL, `name` VARCHAR(100) DEFAULT '' NOT NULL, `label` VARCHAR(255) DEFAULT '' NOT NULL, INDEX `parent` (pid), PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB
```

**Comment by sebastian-meyer** [comment]
@csidirop I made your post into an issue (#1029). Please add all follow-up discussion and information there.

**Pull request merged**